File: 0075-sort-colors/0075-sort-colors.py
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        
        ## initialize three pointers left, right and i
        
        l, r = 0, len(nums) - 1
        i = 0
        
        ## swap function
        
        def swap(i, j):
            temp = nums[i]
            nums[i] = nums[j]
            nums[j] = temp
            
        ## note out i pointer should not cross right pointer
        while i <= r:
            if nums[i] == 0:
                swap(l, i)
                l += 1
                i += 1  ## Increment i here
                
            elif nums[i] == 2:
                swap(i, r)
                r -= 1
                ## note if 2 is encountered we should not increment i
                
            else:
                i += 1  ## Increment i if the element is 1

File: 0075-sort-colors/0075-sort-colors.cpp
class Solution {
public:
    void sortColors(vector<int>& nums) {
        int l = 0, r = nums.size() - 1;
        int i = 0;

        while (i <= r) {
            if (nums[i] == 0) {
                swap(nums[l], nums[i]);
                l++;
                i++;
            } else if (nums[i] == 2) {
                swap(nums[i], nums[r]);
                r--;
            } else {
                i++;
            }
        }
    }
};

/*
int main() {
    // Example input
    std::vector<int> nums = {2, 0, 2, 1, 1, 0};

    Solution solution;
    solution.sortColors(nums);

    std::cout << "Sorted Colors: ";
    for (int num : nums) {
        std::cout << num << " ";
    }
    std::cout << std::endl;

    return 0;
}
*/
